### PR TITLE
Allow get release download files and lfs files with oauth2 token format

### LINKED ISSUE #25257: Cannot access private repo release files through API using bearer token anymore
### Description

Apparently an RC release was pushed to Docker instances using the "lates" tag, so my Gitea instance is now running 1.20.0rc and I cannot downgrade. Since updating, I am no longer able to download files over HTTPS using my token. I am attempting to access the file URL and I am getting 404 errors. If I make the repository public, I am able to download the files just fine.

This bearer token works when listing releases via the API, just not downloading them. Nothing else has changed in my instance.

### Gitea Version

1.20.0+rc0-48-g3afc3e4a7

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Docker

### Database

MySQL

### LINKED ISSUE #26165: Download release asset using api key no longer working
### Description

We are using the api to find out which releases exist for a specific repository. Until 1.19.3 we were able to download those assets as well when passing the api key in the header to the `browser_download_url` which is part of the listing at 
`https://git.fluves.net/api/swagger#/repository/repoGetReleaseByTag`

This no longer works in 1.20.1 (not checked versions between 1.19.3 and 1.20.1).

The headers we pass are:

```
headers = {
    "Authorization": f"token {gitea_token}",
    "Content-Type": "application/json",
    "Accept": "application/json",
}
```

this works for the api, but no longer for the download. The URL given works if I use it from a browser where I logged in manually.

### Gitea Version

1.20.1

### Can you reproduce the bug on the Gitea demo site?

Yes:

Building an example at try.gitea.org

I generated a new token with read:package and read:repository rights.
 `c44ec64232370bc06c9f4aa8c2c67e0121ccc72b`
(nothing valuable is in the repo, it is safe to share this).

eg if I go to: 
https://try.gitea.io/api/v1/repos/johanvdw/bug26165/releases/
and then try to open
https://try.gitea.io/johanvdw/bug26165/releases/download/0.1/README.txt

```
curl -X 'GET' \
  'https://try.gitea.io/api/v1/repos/johanvdw/bug26165/releases/tags/0.1?token=c44ec64232370bc06c9f4aa8c2c67e0121ccc72b' \
  -H 'accept: application/json'
```
response:
```
{
  "id": 10386849,
  "tag_name": "0.1",
  "target_commitish": "main",
  "name": "test",
  "body": "test",
  "url": "https://try.gitea.io/api/v1/repos/johanvdw/bug26165/releases/10386849",
  "html_url": "https://try.gitea.io/johanvdw/bug26165/releases/tag/0.1",
  "tarball_url": "https://try.gitea.io/johanvdw/bug26165/archive/0.1.tar.gz",
  "zipball_url": "https://try.gitea.io/johanvdw/bug26165/archive/0.1.zip",
  "draft": false,
  "prerelease": false,
  "created_at": "2023-07-26T13:02:36Z",
  "published_at": "2023-07-26T13:02:36Z",
  "author": {
...
  },
  "assets": [
    {
      "id": 3458,
      "name": "README.txt",
      "size": 53,
      "download_count": 1,
      "created_at": "2023-07-26T13:03:22Z",
      "uuid": "fe39078d-5858-4afb-a71f-115a313f3cb9",
      "browser_download_url": "https://try.gitea.io/johanvdw/bug26165/releases/download/0.1/README.txt"
    }
  ]
}
```
will give me a browser_download_url in assets, but I'm not able to use that to download the file.

```
curl -X 'GET'  https://try.gitea.io/johanvdw/bug26165/releases/download/0.1/README.txt
Not found.
```
similar results when specifying a token (the first variant was definitely working in 1.19).
```
curl -X 'GET'  https://try.gitea.io/johanvdw/bug26165/releases/download/0.1/README.txt -H 'Authorization: token c44ec64232370bc06c9f4aa8c2c67e0121ccc72b'
curl -X 'GET'  https://try.gitea.io/johanvdw/bug26165/releases/download/0.1/README.txt?token=c44ec64232370bc06c9f4aa8c2c67e0121ccc72b
```

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

docker.

### Database

PostgreSQL

### PR DESCRIPTION
**Pull request by lunny**
Fix #26165
Fix #25257

### CODE REVIEW

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
backport?

**Comment by JakobDev** [comment]
Way better than the old PR.

Maybe we should add a Test that checks, if release assets can be downloaded from a private repo with a token, so this don't happen again.

File: services/auth/oauth2.go [hunk 1]
```diff
@@ -126,7 +126,8 @@ func (o *OAuth2) userIDFromToken(tokenSHA string, store DataStore) int64 {
 // If verification is successful returns an existing user object.
 // Returns nil if verification fails.
 func (o *OAuth2) Verify(req *http.Request, w http.ResponseWriter, store DataStore, sess SessionStore) (*user_model.User, error) {
-	if !middleware.IsAPIPath(req) && !isAttachmentDownload(req) && !isAuthenticatedTokenRequest(req) {
+	if !middleware.IsAPIPath(req) && !isAttachmentDownload(req) && !isAuthenticatedTokenRequest(req) &&
```
Comment (wxiaoguang) [inline services/auth/oauth2.go:129]: Could there be a comment for "why these paths can be ignored"?

Comment (wxiaoguang) [reply services/auth/oauth2.go:129]: Could there be a comment for "why these paths can be ignored"?

Comment (lunny) [reply services/auth/oauth2.go:129]: I thinks a comment is very good. And I found some tests failure because this change. I will fix that test first.

**Comment by JakobDev** [comment]
Why was this closed?

**Comment by xiaoxinpro** [comment]
Excuse me, can this PR solve the problem of using access_token to download the release asset? #26165 
I see the header says "oauth2", which has nothing to do with access_token

**Comment by kerwin612** [comment]
I am very much looking forward to its merger and release as soon as possible.

File: services/auth/oauth2.go (hunk 1 above)
Comment (lunny) [reply services/auth/oauth2.go:129]: 3bc8a024d16b1acd164ad644bcaf86cd817b4813

**Comment by lunny** [comment]
It's ready to review again. Sorry for the late.

**Comment by kirbylink** [comment]
last test still fails :(

**Comment by kirbylink** [comment]
Sorry, for posting again, but this PR is really important for me.
Are the tests now okay? Can we merge the PR?

File: services/auth/oauth2.go [hunk 2]
```diff
@@ -125,7 +125,9 @@ func (o *OAuth2) userIDFromToken(ctx context.Context, tokenSHA string, store Dat
 // If verification is successful returns an existing user object.
 // Returns nil if verification fails.
 func (o *OAuth2) Verify(req *http.Request, w http.ResponseWriter, store DataStore, sess SessionStore) (*user_model.User, error) {
-	if !middleware.IsAPIPath(req) && !isAttachmentDownload(req) && !isAuthenticatedTokenRequest(req) {
+	// These paths are not API paths, but we still want to check for tokens because they maybe in the API returned URLs
+	if !middleware.IsAPIPath(req) && !isAttachmentDownload(req) && !isAuthenticatedTokenRequest(req) &&
+		!gitRawReleasePathRe.MatchString(req.URL.Path) {
```
Comment (wxiaoguang) [inline services/auth/oauth2.go:130]: Why not call "isGitRawReleaseOrLFSPath" ?

File: tests/test_utils.go [hunk 3]
```diff
@@ -213,6 +213,18 @@ func PrepareTestEnv(t testing.TB, skip ...int) func() {
...
+	assert.NoError(t, storage.Clean(storage.Attachments))
+
+	s, err := storage.NewStorage(setting.LocalStorageType, &setting.Storage{
+		Path: filepath.Join(filepath.Dir(setting.AppPath), "tests", "testdata", "data", "attachments"),
+	})
+	assert.NoError(t, err)
+	assert.NoError(t, s.IterateObjects("", func(p string, obj storage.Object) error {
+		_, err = storage.Copy(storage.Attachments, p, s, p)
+		return err
+	}))
```
Comment (wxiaoguang) [inline tests/test_utils.go]: Does every test need such preparation "copy"?

For more than 99% tests, this "copy" just wastes time.

Let the test case prepare the data when it really needs it?

**Review by wxiaoguang — APPROVED** [review]

File: tests/test_utils.go (hunk 3 above)
Comment (lunny) [reply tests/test_utils.go]: Done.

**Comment by kirbylink** [comment]
In 1.20.5 it is working again. Thanks :+1:

**Pull request merged**